subsys/bluetooth/controller/ll_sw/ull_llcp.c

```diff
@@ -1001,36 +1001,119 @@ void ull_cp_tx_ack(struct ll_conn *conn, struct node_tx *tx)
 
 void ull_cp_rx(struct ll_conn *conn, struct node_rx_pdu *rx)
 {
+	struct proc_ctx *ctx_l;
+	struct proc_ctx *ctx_r;
 	struct pdu_data *pdu;
-	struct proc_ctx *ctx;
+	bool unexpected_l;
+	bool unexpected_r;
 
 	pdu = (struct pdu_data *)rx->pdu;
 
-	if (!pdu_is_terminate(pdu)) {
-		/*
-		 * Process non LL_TERMINATE_IND PDU's as responses to active
-		 * procedures
-		 */
+	if (pdu_is_terminate(pdu)) {
+		/*  Process LL_TERMINATE_IND PDU's as new procedure */
+		ctx_l = NULL;
+		ctx_r = NULL;
+	} else {
+		/* Query local and remote activity */
+		ctx_l = llcp_lr_peek(conn);
+		ctx_r = llcp_rr_peek(conn);
+	}
 
-		ctx = llcp_lr_peek(conn);
-		if (ctx && (pdu_is_expected(pdu, ctx) || pdu_is_unknown(pdu, ctx) ||
-			    pdu_is_reject(pdu, ctx))) {
-			/* Response on local procedure */
-			llcp_lr_rx(conn, ctx, rx);
-			return;
-		}
+	if (ctx_l) {
+		/* Local active procedure */
 
-		ctx = llcp_rr_peek(conn);
-		if (ctx && (pdu_is_expected(pdu, ctx) || pdu_is_unknown(pdu, ctx) ||
-			    pdu_is_reject(pdu, ctx))) {
-			/* Response on remote procedure */
-			llcp_rr_rx(conn, ctx, rx);
-			return;
+		if (ctx_r) {
+			/* Local active procedure
+			 * Remote active procedure
+			 */
+
+			unexpected_l = !(pdu_is_expected(pdu, ctx_l) ||
```

Comment:
this reverted logic is bit hard to read for me but I guess this is just a matter of taste :-)